File: files/CH-161.py
import pandas as pd

path = "CH-161 Custom Index Column.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="E:H", skiprows=1, nrows=4)

result = input.assign(rn=input.groupby('Product').cumcount() + 1) \
               .pivot(index='rn', columns='Product', values='Date') \
               .reset_index(drop=True).rename_axis(None, axis=1)

print(result.equals(test)) # True

File: files/100-199/161/CH-161.R
library(tidyverse)
library(readxl)

path = "files/CH-161 Custom Index Column.xlsx"
input = read_excel(path, range = "B2:C11")
test  = read_excel(path, range = "E2:H6")

result = input %>%
  mutate(rn = row_number(), .by = Product) %>%
  pivot_wider(names_from = Product, values_from = Date) %>%
  select(-rn)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE